core/apple/src/SourcesApple.kt

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2017-2023 JetBrains s.r.o. and respective authors and developers.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the LICENCE file.
+ */
+
+package kotlinx.io
+
+import kotlinx.cinterop.*
+import platform.Foundation.*
+import platform.darwin.NSInteger
+import platform.darwin.NSUInteger
+import platform.darwin.NSUIntegerVar
+import platform.posix.uint8_tVar
+import kotlin.native.ref.WeakReference
+
+/**
+ * Returns an input stream that reads from this source. Closing the stream will also close this source.
+ *
+ * @sample kotlinx.io.samples.KotlinxIoSamplesApple.asStream
+ */
+public fun Source.asNSInputStream(): NSInputStream = SourceNSInputStream(this)
+
+@OptIn(InternalIoApi::class, UnsafeNumber::class)
+private class SourceNSInputStream(
+    private val source: Source
+) : NSInputStream(NSData()), NSStreamDelegateProtocol {
+
+    private val isClosed: () -> Boolean = when (source) {
+        is RealSource -> source::closed
+        is Buffer -> {
+            { false }
+        }
+    }
+
+    private var status = NSStreamStatusNotOpen
+    private var error: NSError? = null
+        set(value) {
+            status = NSStreamStatusError
+            field = value
+            source.close()
+        }
+
+    override fun streamStatus() = if (status != NSStreamStatusError && isClosed()) NSStreamStatusClosed else status
+
+    override fun streamError() = error
+
+    override fun open() {
+        if (status == NSStreamStatusNotOpen) {
+            status = NSStreamStatusOpening
+            status = NSStreamStatusOpen
+            postEvent(NSStreamEventOpenCompleted)
+            checkBytes()
+        }
+    }
+
+    override fun close() {
+        if (status == NSStreamStatusError || status == NSStreamStatusNotOpen) return
+        status = NSStreamStatusClosed
+        runLoop = null
+        runLoopModes = listOf()
+        source.close()
+    }
+
+    override fun read(buffer: CPointer<uint8_tVar>?, maxLength: NSUInteger): NSInteger {
+        if (streamStatus != NSStreamStatusOpen && streamStatus != NSStreamStatusAtEnd || buffer == null) return -1
+        status = NSStreamStatusReading
+        try {
+            if (source.exhausted()) {
+                status = NSStreamStatusAtEnd
+                return 0
+            }
+            val toRead = minOf(maxLength.toLong(), source.buffer.size, Int.MAX_VALUE.toLong()).toInt()
+            val read = source.buffer.readAtMostTo(buffer, toRead).convert<NSInteger>()
+            status = if (source.exhausted()) NSStreamStatusAtEnd else NSStreamStatusOpen
```

Comment:
It might be a problem: suppose we have a pair of sockets, client sends some data, we process it and respond, and there will be no more data from the client until the server respond. Here, we'll fetch all the data from a network buffer using `readAtMostTo` and then will call `exhausted` that will hang until connection timed out.
I think it's better to revert a commit changing that behavior and set the status to open after performing a read. The next read will change it to at-end if there are no more data.

Reply:
Ok, I was just comparing to the behavior of `NSInputStream` and it does have the status set to `NSStreamStatusAtEnd` after the last read, at least for the `NSData` implementation. I can see how that doesn't necessarily work for all `Source` implementations though. I'll revert this.